agent/acs/handler/payload_responder.go

```diff
@@ -0,0 +1,289 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package handler
+
+import (
+	"fmt"
+
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	apiappmesh "github.com/aws/amazon-ecs-agent/agent/api/appmesh"
+	apitask "github.com/aws/amazon-ecs-agent/agent/api/task"
+	apitaskstatus "github.com/aws/amazon-ecs-agent/agent/api/task/status"
+	"github.com/aws/amazon-ecs-agent/agent/data"
+	"github.com/aws/amazon-ecs-agent/agent/engine"
+	"github.com/aws/amazon-ecs-agent/agent/eventhandler"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/acs/model/ecsacs"
+	apieni "github.com/aws/amazon-ecs-agent/ecs-agent/api/eni"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/credentials"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	loggerfield "github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	"github.com/aws/aws-sdk-go/aws"
+)
+
+// skipAddTaskComparatorFunc defines the function pointer that accepts task status
+// and returns the boolean comparison result.
+type skipAddTaskComparatorFunc func(apitaskstatus.TaskStatus) bool
+
+// taskMessageHandler implements TaskMessageHandler.
+type taskMessageHandler struct {
+	taskEngine                  engine.TaskEngine
+	ecsClient                   api.ECSClient
+	dataClient                  data.Client
+	taskHandler                 *eventhandler.TaskHandler
+	credentialsManager          credentials.Manager
+	latestSeqNumberTaskManifest *int64
+}
+
+func (tmHandler *taskMessageHandler) ProcessTaskMessage(message *ecsacs.PayloadMessage,
+	ackFunc func(*ecsacs.AckRequest, []*ecsacs.IAMRoleCredentialsAckRequest)) error {
+
+	credentialsAcks, allTasksHandled := tmHandler.addPayloadTasks(message)
+
+	// Update latestSeqNumberTaskManifest for it to get updated in state file.
```

Comment:
Not related to PR itself but for my knowledge - how does the acs handler decide the order of handling messages? Would it be possible that it handles a message with larger seqNum first and then get back to lower ones? When this happens and we do need to recover from state file, would it (and how would it) affect the message handling?

Reply:
Although unrelated to the PR, these are good questions! I'll take a deeper look into this and get back to you.

Reply:
> how does the acs handler decide the order of handling messages? 

Session/ACS handler consumes messages in the order that they are received.

> Would it be possible that it handles a message with larger seqNum first and then get back to lower ones? When this happens and we do need to recover from state file, would it (and how would it) affect the message handling?

Agent keeps track of the highest sequence number received in both messages of type`PayloadMessage` and `TaskManifestMessage`. If the sequence number on a message with type `TaskManifestMessage` is lower than the latest known one, the the task manifest message is considered stale and discarded.

ACS guarantees that the sequence number is increasing and valid.

Reply:
Thanks for the research!